### PR TITLE
Metadata object is share across different clusters and hosts

### PR DESCRIPTION
**Pull request by zyfjeff**
Signed-off-by: tianqian.zyf <tianqian.zyf@alibaba-inc.com>

Currently, our host metadata use `map<string, google.protobuf.struct >` type, because of the use of `google.protobuf.struct`, which contains a lot of type reflection information, which takes up more memory space. Even without any metadata information, an empty object would take up 144 bytes.

This PR does the following two optimizations:

* Metadata objects are shared across clusters and hosts
* Optimize the empty value for existing metadata

Description:
Risk Level: low
Testing: existing tests
Docs Changes: N/A
Release Notes: N/A
[Optional Fixes #Issue] #9532 
[Optional Deprecated:]

### CODE REVIEW

**Comment by zyfjeff** [comment]
The results of the actual test were excellent. The 100W host did not come with any metadata information, occupying about 1.3g, about 1.9g with metadata, and reduced to the only 960M after optimization

**Comment by zyfjeff** [comment]
The core implementation is `ObjectSharedPool`, and the rest of the code is mostly interface changes.

File: source/common/upstream/upstream_impl.h [hunk 1]
```diff
@@ -91,13 +91,13 @@ class HostDescriptionImpl : virtual public HostDescription,
...
-  const std::shared_ptr<envoy::config::core::v3::Metadata> metadata() const override {
+  MetadataConstSharedPtr metadata() const override {
     absl::ReaderMutexLock lock(&metadata_mutex_);
     return metadata_;
   }
-  void metadata(const envoy::config::core::v3::Metadata& new_metadata) override {
+  void metadata(MetadataConstSharedPtr new_metadata) override {
     absl::WriterMutexLock lock(&metadata_mutex_);
-    metadata_ = std::make_shared<envoy::config::core::v3::Metadata>(new_metadata);
+    metadata_ = new_metadata;
```
Comment (zuercher) [inline source/common/upstream/upstream_impl.h]: Potential threading problem here. The design requires that `metadata_` be destroyed by the main thread. I think that's the common case because normally the shared Host pointer is only destroyed when all referencing load balancers have been destroyed.

However, the following scenario seems possible:
1. a worker thread invokes `metadata()` and retrieves a MetadataConstSharedPtr (e.g. for header formatting).
2. before that shared ptr is released, an EDS update on the main thread invokes this method

I think that the shared_ptr in 1 could end up being the last reference, and would then be destroyed on the worker thread.

File: test/common/shared_pool/shared_pool_test.cc [hunk 2]
```diff
@@ -23,27 +27,47 @@ TEST(SharedPoolTest, Basic) {
...
 
 TEST(SharedPoolTest, NonThreadSafeForGetObjectDeathTest) {
-  std::shared_ptr<ObjectSharedPool<int>> pool;
-
-  std::thread another_thread([&pool] { pool = std::make_shared<ObjectSharedPool<int>>(); });
-  another_thread.join();
+  Event::MockDispatcher dispatcher;
+  auto pool = std::make_shared<ObjectSharedPool<int>>(dispatcher);
+  pool->thread_id_ = std::thread::id();
+#ifdef NDEBUG
```
Comment (zuercher) [inline test/common/shared_pool/shared_pool_test.cc]: Why is this necessary?

File: source/common/upstream/upstream_impl.cc [hunk 3]
```diff
@@ -1342,12 +1343,16 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
...
 
       // Did metadata change?
-      bool metadata_changed = false;
+      bool metadata_changed = true;
       if (host->metadata() && existing_host->second->metadata()) {
         metadata_changed = !Protobuf::util::MessageDifferencer::Equivalent(
             *host->metadata(), *existing_host->second->metadata());
       }
 
+      if (!host->metadata() && !existing_host->second->metadata()) {
```
Comment (zuercher) [inline source/common/upstream/upstream_impl.cc]: nit: else if

File: source/common/shared_pool/shared_pool.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  std::size_t poolSize() const {
+    ASSERT(std::this_thread::get_id() == thread_id_);
+    return object_pool_.size();
+  }
+
+private:
+  std::thread::id thread_id_;
+  absl::flat_hash_map<size_t, std::weak_ptr<T>> object_pool_;
+  Event::Dispatcher& dispatcher_;
+  FRIEND_TEST(SharedPoolTest, NonThreadSafeForGetObjectDeathTest);
```
Comment (asraa) [inline source/common/shared_pool/shared_pool.h]: FWIW `FRIEND_TEST` is only needed to manipulate the `thread_id_`. Would it be possible to not use the `FRIEND_TEST` construct and use a `setThreadIdForTest()`?

Comment (asraa) [reply source/common/shared_pool/shared_pool.h]: Or possibly `threadFactoryForTest().createThread`

File: test/common/shared_pool/shared_pool_test.cc (hunk 2 above)
Comment (zyfjeff) [reply test/common/shared_pool/shared_pool_test.cc]: release conditions will not trigger the assertions, even if not thread-safe can still run, this time will trigger the object's destructor to call `deleteObject` method, and then the internal call `disptaher: : post` method.

Comment (zyfjeff) [reply test/common/shared_pool/shared_pool_test.cc]: The debug condition triggers the assertion directly, not the object destructor, so the call to `dispatcher::post` is not triggered.

File: source/common/shared_pool/shared_pool.h (hunk 4 above)
Comment (zyfjeff) [reply source/common/shared_pool/shared_pool.h]: `setThreadIdForTest` is more suitable, the way to create threads is still not flexible enough to test

**Comment by zyfjeff** [comment]
@asraa  @zuercher ping you

File: source/common/shared_pool/shared_pool.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+};
+
+using MetadataConstSharedPtr = std::shared_ptr<const envoy::config::core::v3::Metadata>;
+
+using ConstMetadataSharedPoolSharedPtr =
+    std::shared_ptr<ObjectSharedPool<const envoy::config::core::v3::Metadata, MessageUtil>>;
+
+ConstMetadataSharedPoolSharedPtr getConstMetadataSharedPool(Singleton::Manager& manager);
+
+} // namespace Envoy
```
[the remark this answers is not in the document]
Comment (zuercher) [reply source/common/shared_pool/shared_pool.h]: These came back.

[the remark this answers is not in the document]
Comment (zuercher) [reply source/common/shared_pool/shared_pool.h]: Nevermind. Github is lying to me.

File: test/common/shared_pool/shared_pool_test.cc (hunk 2 above)
Comment (zuercher) [reply test/common/shared_pool/shared_pool_test.cc]: Let me see if I understand. In a release build, the ASSERT calls are no-ops and so in this test the getObject call succeeds and then the ObjectSharedPool is destroyed, which deletes the object.

In that case, maybe make the entire test `#ifndef NDEBUG` with an explanation of why it's not run in release?

Comment (zyfjeff) [reply test/common/shared_pool/shared_pool_test.cc]: yup, your understanding is correct,  I'll add some comments to make the code easier to understand

**Comment by zyfjeff** [comment]
@zuercher @asraa Is there any other issue for this pr to review?

File: source/common/shared_pool/shared_pool.h [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace SharedPool {
+
+/**
+ * Used to share objects that have the same content.
+ * control the life cycle of shared objects by reference counting
+ * Note: this class has only deleteObject for thread safety, and the other member methods are
+ * non-thread safe
```
Comment (mattklein123) [inline source/common/shared_pool/shared_pool.h]: Do you mind adding a bit more comments here in the ownership model? And possibly more comments below? This is neat and makes sense to me but it took a few read throughs to understand which thread is doing what, the use of weak/strong pointers, etc.

File: source/common/shared_pool/shared_pool.h [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+      auto lock_object = object_it->second.lock();
+      if (lock_object) {
+        return lock_object;
+      }
+    }
+
+    auto this_shared_ptr = this->shared_from_this();
+    std::shared_ptr<T> obj_shared(new T(obj), [hashed_value, this_shared_ptr](T* ptr) {
+      this_shared_ptr->deleteObject(hashed_value);
+      delete ptr;
```
Comment (mattklein123) [inline source/common/shared_pool/shared_pool.h]: I *think* this can race due to the same hash value. Meaning, you delete the pointer here, but then you lock the weak pointer on the main thread and then you would cash with no internal value? I think you need to actually do the deletion on the main thread? Can you fix if necessary and add more comments/tests around this?

File: source/common/shared_pool/shared_pool.h (hunk 6 above)
Comment (zyfjeff) [reply source/common/shared_pool/shared_pool.h]: I add some comments, but I don't know if it's enough for someone else to read.

File: source/common/shared_pool/shared_pool.h (hunk 7 above)
Comment (zyfjeff) [reply source/common/shared_pool/shared_pool.h]: I don't think there will be a race condition, because the object destructor and the object obtained through the weak_ptr are thread-safe, and when the object destructor is not finished, weak_ptr cannot get the real object through lock.

But I found another problem when an object is destructed, if it is not in the main thread, it will post a task to the main thread, and if the main thread gets an object with the same hash value through `getObject` at this time, it will cause a new weak_ptr to be stored in the map, and then execute `deleteObject` will delete the new weak_ptr.  I have fixed this issue and added the related unit test.

Comment (zyfjeff) [reply source/common/shared_pool/shared_pool.h]: https://en.cppreference.com/w/cpp/memory/weak_ptr/expired

> Equivalent to use_count() == 0. The destructor for the managed object may not yet have been called, but this object's destruction is imminent (or may have already happened).

Comment (mattklein123) [reply source/common/shared_pool/shared_pool.h]: Can you add more comments here? I see that you switched the order of delete and delete object?

File: source/common/shared_pool/shared_pool.h [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    return obj_shared;
+  }
+
+  std::size_t poolSize() const {
+    ASSERT(std::this_thread::get_id() == thread_id_);
+    return object_pool_.size();
+  }
+
+  void setThreadIdForTest(const std::thread::id& thread_id) { thread_id_ = thread_id; }
```
Comment (mattklein123) [inline source/common/shared_pool/shared_pool.h]: I've been looking at this for 20 minutes. Although I think this is correct, this is complicated enough that I think we need real tests with threads. Can you please remove this function and actually have the tests use real threads? You can use my ThreadSynchronizer class to relatively easy invoke all of the scenarios that you want to test.

File: source/common/shared_pool/shared_pool.h (hunk 7 above)
Comment (zyfjeff) [reply source/common/shared_pool/shared_pool.h]: There is no particular intention, just hope to release the ptr as soon as possible, which can avoid the ptr being exposed by mistake, resulting in some undefined behavior.

File: source/common/shared_pool/shared_pool.h (hunk 8 above)
Comment (zyfjeff) [reply source/common/shared_pool/shared_pool.h]: The `setThreadIdForTest` function isn't ready to be deleted yet, and I need to use this function to modify the thread id because I'm using Mock's `Dispathcer`, which doesn't help me deliver deleteObject to the corresponding thread. So I still need to switch different thread ids through setThreadIdForTest

Comment (zyfjeff) [reply source/common/shared_pool/shared_pool.h]: All changed to real  test with threads,  and remove setThreadIdForTest

File: source/common/shared_pool/shared_pool.h [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+  /**
+   * @return a thread synchronizer object used for reproducing a race-condition in tests.
+   */
+  Thread::ThreadSynchronizer& sync() { return sync_; }
+  static const char DeleteObjectOnMainThread[];
+  static const char ObjectDeleterEntry[];
+
+private:
+  std::thread::id thread_id_;
```
Comment (mattklein123) [inline source/common/shared_pool/shared_pool.h]: nit: const

File: source/common/shared_pool/shared_pool.h [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+ */
+template <typename T, typename HashFunc = std::hash<T>,
+          class = typename std::enable_if<std::is_copy_constructible<T>::value>::type>
+class ObjectSharedPool : public Singleton::Instance,
+                         public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc>>,
+                         NonCopyable {
+public:
+  ObjectSharedPool(Event::Dispatcher& dispatcher)
+      : thread_id_(std::this_thread::get_id()), dispatcher_(dispatcher) {}
+  ~ObjectSharedPool() = default;
```
Comment (mattklein123) [inline source/common/shared_pool/shared_pool.h]: nit: delete

File: test/common/shared_pool/shared_pool_test.cc [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+  Api::ApiPtr api_;
+  Event::DispatcherPtr dispatcher_;
+  Thread::ThreadPtr dispatcher_thread_;
+  Event::TimerPtr keepalive_timer_;
+  absl::Notification go_;
+};
+
+TEST_F(SharedPoolTest, Basic) {
+  Event::MockDispatcher dispatcher;
```
Comment (mattklein123) [inline test/common/shared_pool/shared_pool_test.cc]: Can we just use the real dispatcher/pool for all tests in this file? It will make it much more likely to catch issues with TSAN, etc.

**Comment by zyfjeff** [comment]
`sudo ./ci/run_envoy_docker.sh ./ci/do_ci.sh bazel.tsan //test/common/shared_pool/... --runs_per_test=100`

```
Target //test/common/shared_pool:shared_pool_test up-to-date:
  bazel-bin/test/common/shared_pool/shared_pool_test
INFO: Elapsed time: 82.990s, Critical Path: 26.80s
INFO: 100 processes: 100 processwrapper-sandbox.
INFO: Build completed successfully, 101 total actions
//test/common/shared_pool:shared_pool_test                               PASSED in 2.6s
  Stats over 100 runs: max = 2.6s, min = 1.6s, avg = 2.4s, dev = 0.2s
```

**Comment by zyfjeff** [comment]
> Once all of the tests use the real dispatcher and real threads, can you make sure to test a TSAN build with `runs_per_test=1000` to check for flakes? Thank you!

Thanks a lot, `tsan` did find several thread safety issues in the test code.

**Comment by zyfjeff** [comment]
`sudo ./ci/run_envoy_docker.sh ./ci/do_ci.sh bazel.tsan //test/common/shared_pool/... --runs_per_test=1000`

```
Target //test/common/shared_pool:shared_pool_test up-to-date:
  bazel-bin/test/common/shared_pool/shared_pool_test
INFO: Elapsed time: 162.931s, Critical Path: 5.18s
INFO: 1000 processes: 1000 processwrapper-sandbox.
INFO: Build completed successfully, 1001 total actions
//test/common/shared_pool:shared_pool_test                               PASSED in 3.4s
  Stats over 1000 runs: max = 3.4s, min = 1.9s, avg = 2.4s, dev = 0.1s
```

File: source/common/shared_pool/shared_pool.h [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+      // There may be new inserts with the same hash value before deleting the old element,
+      // so there is no need to delete it at this time.
+      if (object_pool_.find(hash_key) != object_pool_.end() &&
+          object_pool_[hash_key].use_count() == 0) {
+        object_pool_.erase(hash_key);
+      }
+    } else {
+      // Most of the time, the object's destructor occurs in the main thread, but with some
+      // exceptions, it is destructed in the worker thread. In order to keep the object_pool_ thread
+      // safe, the deleteObject needs to be delivered to the main thread.
```
Comment (htuch) [inline source/common/shared_pool/shared_pool.h]: What are these exceptions @zyfjeff? It seems pretty dangerous to allow objects created on one thread (main) be destructed on another (worker).

File: source/common/shared_pool/shared_pool.h (hunk 7 above)
Comment (htuch) [reply source/common/shared_pool/shared_pool.h]: Who would be accessing ptr? This seems pretty unclear to me. I think the comment should be further clarified with an example, as it's unclear what the concrete danger is here.

File: source/common/shared_pool/shared_pool.h [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+      // Used for testing to simulate some race condition scenarios
+      sync_.syncPoint(DeleteObjectOnMainThread);
+      dispatcher_.post([hash_key, this_shared_ptr] { this_shared_ptr->deleteObject(hash_key); });
+    }
+  }
+
+  std::shared_ptr<T> getObject(const T& obj) {
+    ASSERT(std::this_thread::get_id() == thread_id_);
+    auto hashed_value = HashFunc{}(obj);
+    auto object_it = object_pool_.find(hashed_value);
```
Comment (htuch) [inline source/common/shared_pool/shared_pool.h]: Is hashing sufficient here (in the absence of a crypto strong hash)? Should we have hash followed by an equality comparison?

File: source/common/shared_pool/shared_pool.h [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+    auto this_shared_ptr = this->shared_from_this();
+    std::shared_ptr<T> obj_shared(new T(obj), [hashed_value, this_shared_ptr](T* ptr) {
+      this_shared_ptr->sync().syncPoint(ObjectSharedPool<T>::ObjectDeleterEntry);
+      // release ptr as early as possible to avoid exposure of ptr, resulting in undefined behavior.
+      delete ptr;
+      this_shared_ptr->deleteObject(hashed_value);
+    });
+
+    // When inserted, it is possible that the old elements still exist before they can be deleted,
```
Comment (htuch) [inline source/common/shared_pool/shared_pool.h]: Can you give an example sequence here that demonstrates?

File: source/common/shared_pool/shared_pool.h (hunk 12 above)
Comment (zyfjeff) [reply source/common/shared_pool/shared_pool.h]: https://github.com/envoyproxy/envoy/pull/10042#discussion_r379155904
 Some scenarios are mentioned in this comments, and I'm here to ensure thread safety by `dispatcher::post` converting these scenarios that are destructed in work threads into main thread.

File: source/common/shared_pool/shared_pool.h (hunk 14 above)
Comment (zyfjeff) [reply source/common/shared_pool/shared_pool.h]: https://github.com/envoyproxy/envoy/blob/45d86449ff28395b3e9edf2db1b001e633f21190/test/common/shared_pool/shared_pool_test.cc#L120

 I don't know if this test case is enough for you to understand?

File: source/common/shared_pool/shared_pool.h (hunk 13 above)
Comment (zyfjeff) [reply source/common/shared_pool/shared_pool.h]: Yes, there is a need to consider hash conflicts. let me open another PR to perfect here.

File: source/common/shared_pool/shared_pool.h (hunk 7 above)
Comment (zyfjeff) [reply source/common/shared_pool/shared_pool.h]: example:

```
    std::shared_ptr<T> obj_shared(new T(obj), [hashed_value, this_shared_ptr](T* ptr) {
      this_shared_ptr->sync().syncPoint(ObjectSharedPool<T>::ObjectDeleterEntry);
       // The error passes the ptr to a function that may not be executed immediately.
       delete ptr;
    });
```

**Comment by zyfjeff** [comment]
@htuch Thank you for your valuable comments

**Comment by zyfjeff** [comment]
@htuch  I am sorry for the late reply, Github don't seem to have informed me

File: source/common/shared_pool/shared_pool.h (hunk 12 above)
Comment (htuch) [reply source/common/shared_pool/shared_pool.h]: I see; I think it would be helpful to have that example (of EDS cleanup on the main thread) in this comment block, otherwise the reasoning gets lost. Seems reasonable to destruct on the main thread as you do here.

File: source/common/shared_pool/shared_pool.h (hunk 14 above)
Comment (htuch) [reply source/common/shared_pool/shared_pool.h]: I'm not grokking it. I think putting in code comments a better explanation/example would be helpful, thanks.

File: source/common/shared_pool/shared_pool.h (hunk 13 above)
Comment (jmarantz) [reply source/common/shared_pool/shared_pool.h:64]: Is this follow-up going to occur? The current code 18 months later is still doing a hash only with no equality comparison.

More concerning is that I see a lot of tsan errors, at least with the toolchain on my workstation. I'm not sure if this is toolchain or hardware related or if something regressed since the original test. I noticed also that docker was used for the original tsan tests and I'm not sure if that makes a difference. I tend to run my tests on RBE without docker these days:
```
bazel test --config=libc++ --config=remote  //test/common/shared_pool:shared_pool_test --config=clang-tsan
...
//test/common/shared_pool:shared_pool_test                               FAILED in 990 out of 1000 in 9.6s
```
I do note that this test passes 1000/1000 times for me when I don't use tsan.

I note also that https://stackoverflow.com/questions/67771015/shared-from-this-and-thread-safety appears conflicted about whether shared_from_this is thread-safe, and it points to https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr#Notes which doesn't appear to address the issue specifically on a quick skim.

Comment (jmarantz) [reply source/common/shared_pool/shared_pool.h:64]: I'm assuming also that CI is not finding this tsan error generally though I am seeing flakes in coverage tests.

https://gist.github.com/jmarantz/7657d187ad7b011a12b5878be64bc209 is the log of one of my tsan failures.

Comment (jmarantz) [reply source/common/shared_pool/shared_pool.h:64]: I have a resolution for the tsan test failure -- i think the issue was just that the test thread was racing against the main thread for the destruction of `o2` in 2 tests. https://github.com/envoyproxy/envoy/pull/18262

Comment (jmarantz) [reply source/common/shared_pool/shared_pool.h:64]: I opened https://github.com/envoyproxy/envoy/issues/24617 to reflect that a follow-up is needed to make this implementation correct in presence of possible hash collisions.
